cpp/oai_pmh_harvester.cc

```diff
@@ -170,85 +157,179 @@ bool ListRecords(const std::string &url, const unsigned time_limit_in_seconds_pe
+                    new_record.insertField(MARC::Tag("655"), 'a', last_data, ' ', '7');
+                else if (tag == "source")
+                    new_record.insertField(MARC::Tag("786"), 'n', last_data, '0', ' ');
+                else if (tag == "coverage")
+                    new_record.insertField(MARC::Tag("500"), 'a', last_data);
+                else if (tag == "subject")
+                    new_record.insertField(MARC::Tag("653"), 'a', last_data);
+                else
+                    LOG_ERROR("Unhandled tag: \"" + xml_part.data_ + "\"!");
+            } else if (xml_part.data_ == "oai_dc:dc") {
+                if (WriteIfNotDupe(new_record, dups_db, marc_writer))
+                    counter++;
+                last_data.clear();
+                break;
             }
-            dups_db->addOrReplace(checksum, TimeUtil::GetCurrentDateAndTime());
         }
+    }
+
+    LOG_INFO("Generated " + std::to_string(counter) + " MARC record(s).");
+}
+
+
+void GenerateValidatedOutputFromMARC(KeyValueDB * const dups_db, MARC::Reader * const marc_reader,
+                                     const std::string &control_number_prefix, MARC::Writer * const marc_writer)
+{
+    unsigned record_number(0), counter(0);
+    while (MARC::Record record = marc_reader->read()) {
+        if (not record.hasValidLeader())
+            continue;
+        ++record_number;
 
         if (record.getControlNumber().empty()) {
-            const std::string control_number(control_number_prefix + StringUtil::Map(StringUtil::ToString(++counter, 10, 10), ' ', '0'));
+            const std::string control_number(control_number_prefix
+                                             + StringUtil::Map(StringUtil::ToString(record_number, 10, 10), ' ', '0'));
             record.insertField("001", control_number);
         }
 
-        marc_writer->write(record);
+        if (WriteIfNotDupe(record, dups_db, marc_writer))
+            counter++;
     }
+
+    LOG_INFO("Extracted " + std::to_string(counter) + " MARC record(s).");
 }
 
 
 } // unnamed namespace
 
 
 int Main(int argc, char **argv) {
-    std::unique_ptr<KeyValueDB> dups_db;
-    if (argc > 1 and std::strcmp(argv[1], "--skip-dups") == 0) {
-        dups_db = CreateOrOpenKeyValueDB();
+    if (argc > 1 and std::strcmp(argv[1], "--skip-dups") == 0)
         --argc, ++argv;
-    }
 
     bool ignore_ssl_certificates(false);
     if (argc > 1 and std::strcmp(argv[1], "--ignore-ssl-certificates") == 0) {
        ignore_ssl_certificates = true;
        --argc, ++argv;
     }
 
-    if (argc != 6 and argc != 7)
-        Usage();
+    if (argc != 7 and argc != 8)
+        ::Usage("[--skip-dups] [--ignore-ssl-certificates] base_url metadata_prefix [harvest_set_or_identifier] "
+                "control_number_prefix output_filename time_limit_per_request path_to_dups_database\n"
+                "If \"--skip-dups\" has been specified, records that we already encountered in the past won't\n"
```

Comment:
Do we really need separate args `[--skip-dups]` and `path_to_dups_database`?
It might be easier if we merge these together as [--path-to-dups-database=<path>], which implies that dups should be checked if the path is given.

Reply:
That sounds like a great idea! => Done!